in-progress/8754-slashing-staking/design.md

```diff
@@ -0,0 +1,436 @@
+
+- `VALIDATING` meaning that it is currently active,
+- `LIVING` being that it is not participating in block production but also not exiting the system currently. It is possible to land in this state if a slash cause the validator to fall below the `MINIMUM_STAKE`.
+- `EXITING` the operator is currently waiting to exit or yet to claim the funds
+
+The `withdrawer` is a separate actor, specified per operator that is able to initialise a removal of the operator from the set - without the help of the `proposer` or `attester`. The `withdrawer` can initiate a withdrawal of the `stake` to some `recipient`, moving the status from `VALIDATING|LIVING` to `EXITING`, removing the operator from the set and start the delay. After the delay, the funds can be relayed to the specified `recipient`. This ensure that it is possible to keep the "account in control" in cold store or be a contract itself - lending itself well to supporting LST's
+
+We allow a `SLASHER` to reduce the `stake` of an operator, and updating its state to `LIVING` and removing it from the set if `stake < MINIMUM_STAKE`.
+
+The `SLASHER` can then be implemented as a variant of the `Governance.sol` contract, that skips the wider vote and allow direct execution of the payload if the validators agree. In this manner, the node software itself can be responsible for detecting misbehaviour and submitting slashing proposals, giving us a lot of flexibility as it is mainly off-chain code.
+
+> [!info] Top-ups  
+> We are not supporting "topping up" the account, since it introduce a few additional code paths that we simple are not interested in for this minimal setup.
+
+## Users and Interaction
+
+The users of the staking contract are:
+
+- The rollup instance, to sample from the set
+- The slasher, to slash operators
+- Actors that wish to add another operator to the set, exit or vote on slashing
+
+The implementation itself can be made as part of the rollup instance, or as a separate contract, for now we are just going with the same contract as it makes some of the data access slightly simpler as we don't need to provide as many getters.
+
+For users indending to be validators, there is mainly three points of interest:
+
+- Joining the set
+- Exiting the set
+- Voting on slashing
+
+Joining and exiting is done directly on the contract, likely using a frontend or CLI but connected to a cold-storage wallet. These actions are seldom and deal with value so it is important to keep it safe.
+
+As mentioned, the voting on slashing is to be done by the node itself by "slashing-modules". Essentially a small pieces of code that the node is running that will identify misbehaviour and submit or vote on slashing proposal. Votes are collected in the same manner as the `GovernanceProposer` used for governance proposals, with the caveat that if it is "to be proposed" it will go to the slasher where it becomes executable - it does not need to pass a normal governance vote.
```

Comment:
```suggestion
As mentioned, the voting on slashing is to be done by the node itself by "slashing-modules". Essentially a small pieces of code that the node is running that will identify misbehaviour and submit or vote on slashing proposal. Votes are collected in the same manner as the `GovernanceProposer` is used for governance proposals, with the caveat that if it is "to be proposed" it will go to the slasher where it becomes executable - it does not need to pass a normal governance vote.
```